tendermint/src/lite/error.rs

```diff
@@ -6,63 +6,74 @@ use std::time::{SystemTime, SystemTimeError};
 
 /// The main error type verification methods will return.
-/// See [`ErrorKind`] for the different kind of errors.
+/// See [`Kind`] for the different kind of errors.
 pub type Error = anomaly::Error<Kind>;
 
+/// All error kinds related to the light client.
 #[derive(Clone, Debug, Error)]
 pub enum Kind {
+    /// The provided header expired.
     #[error("old header has expired at {at:?} (now: {now:?})")]
     Expired { at: SystemTime, now: SystemTime },
 
+    /// Trusted header is from the future.
     #[error("duration error {:?}", _0)]
     DurationOutOfRange(#[from] SystemTimeError),
 
+    /// Header height smaller than expected.
     #[error("expected height >= {expected} (got: {got})")]
     NonIncreasingHeight { got: u64, expected: u64 },
 
-    #[error("could not verify signature")]
-    InvalidSignature, // TODO: deduplicate with tendermint::ErrorKind::SignatureInvalid
-
+    /// Invalid validator hash.
     #[error("header's validator hash does not match actual validator hash ({header_val_hash:?}!={val_hash:?})")]
     InvalidValidatorSet {
         header_val_hash: Hash,
         val_hash: Hash,
     },
 
+    /// Invalid next validator hash.
     #[error("header's next validator hash does not match next_val_hash ({header_next_val_hash:?}!={next_val_hash:?})")]
     InvalidNextValidatorSet {
         header_next_val_hash: Hash,
         next_val_hash: Hash,
     },
 
+    /// Commit is not for the header we expected.
     #[error(
         "header hash does not match the hash in the commit ({header_hash:?}!={commit_hash:?})"
     )]
     InvalidCommitValue {
         header_hash: Hash,
         commit_hash: Hash,
-    }, // commit is not for the header we expected
-
-    #[error("error validating commit signatures: {info}")]
-    InvalidCommitSignatures { info: String }, // Note: this is only used by implementation (ie. expected return in Commit::validate())
+    },
 
+    /// Signed power does not account for +2/3 of total voting power.
     #[error("signed voting power ({signed}) do not account for +2/3 of the total voting power: ({total})")]
     InvalidCommit { total: u64, signed: u64 },
 
+    /// This means the trust threshold (default +1/3) is not met.
     #[error("signed voting power ({}) is too small fraction of total voting power: ({}), threshold: {}",
         .signed, .total, .trust_treshold
     )]
     InsufficientVotingPower {
         total: u64,
         signed: u64,
         trust_treshold: String,
-    }, // trust threshold (default +1/3) is not met
-
-    #[error("{0:?}")]
-    RequestFailed(String), // Note: we don't want to depend on the tendermint rpc error types, otherwise we should do #[from] Error instead of expecting a string
+    },
 
+    /// This is returned if an invalid TrustThreshold is created.
     #[error("A valid threshold is `1/3 <= threshold <= 1`, got: {got}")]
     InvalidTrustThreshold { got: String },
+
+    /// Use the [`Kind::context`] method to wrap the underlying error of
```

Comment:
How does the access of the context work from a callers perspective, given they encountered a wrapped `RequestFailed`? Asking mostly because I'm not familiar with the APIs yet.

Reply:
It's type erased and stored as a `Box<std::error::Error>`, and available as the [`Error::source`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source).

If you just want a string representation, it impls `Display`, otherwise if you'd like the concrete type again you need to use [`Error::downcast_ref`](https://doc.rust-lang.org/std/error/trait.Error.html#method.downcast_ref)

Reply:
@xla I've added some small tests using source to explore and document how this looks like:
https://github.com/interchainio/tendermint-rs/blob/52e21325983f445fc87bdb7c888e7ae1859275bd/tendermint/src/lite/error.rs#L91-L127